Sources/SecureXPC/XPCRequestContext.swift

```diff
@@ -0,0 +1,93 @@
+//
+//  XPCRequestContext.swift
+//  SecureXPC
+//
+//  Created by Josh Kaplan on 2022-02-22.
+//
+
+import Foundation
+
+/// Provides information about the current request being handled by an ``XPCServer``.
+///
+/// Accessing the properties exposed by this class is a programming error unless called from inside of a closure registered with and called by an `XPCServer`.
+///
+/// ## Topics
+/// ### Client Information
+/// - ``clientCode``
+/// - ``effectiveUserId``
+/// - ``effectiveGroupId``
+public class XPCRequestContext {
+    private let connection: xpc_connection_t
+    private let message: xpc_object_t
+    
+    private init(connection: xpc_connection_t, message: xpc_object_t) {
+        self.connection = connection
+        self.message = message
+    }
+    
+    // MARK: task local
+        
+    @available(macOS 10.15.0, *)
+    @TaskLocal
+    private static var currentForTask: XPCRequestContext?
+    
+    @available(macOS 10.15.0, *)
+    @discardableResult internal static func setForTask<Success, Failure>(
+        connection: xpc_connection_t,
+        message: xpc_object_t,
+        operation: () throws -> Task<Success, Failure>
+    ) rethrows -> Task<Success, Failure> {
+        try $currentForTask.withValue(XPCRequestContext(connection: connection, message: message)) {
+            try operation()
+        }
+    }
+    
+    // MARK: thread local
+    
+    private static let contextKey = UUID()
+    
+    @discardableResult internal static func setForCurrentThread<R>(
+        connection: xpc_connection_t,
+        message: xpc_object_t,
+        operation: () throws -> R
+    ) rethrows -> R {
+        Thread.current.threadDictionary[contextKey] = XPCRequestContext(connection: connection, message: message)
+        let result = try operation()
+        Thread.current.threadDictionary.removeObject(forKey: contextKey)
+        
+        return result
+    }
```

Comment:
I would extract this concern into your own property wrapper, which automatically picks between task local or thread local storage. IDK what I would call it though, naming is hard.

Reply:
How would I be able to automatically determine which type of storage should be used?

Reply:
I'm going to merge this, but happy to revisit this once I understand what you're suggesting.

Reply:
Using the same `@available(macOS 10.15.0, *)`  (or but in the `if #available ...` runtime variant) checks that you already use here.

Reply:
But setting a synchronous error handler is still possible (and perfectly reasonable) to use on macOS 10.15 or later.

Right now the caller explicitly calls either `setForTask(...)` or `setForCurrentThread(...)` which lets me know which type of storage needs to be used. I've been interpreting your comment as being about automatically determining whether the storage needs to be per task or per thread, but without an explicit decision I don't know how to determine that.